resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -481,20 +478,29 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
   }
 
   private def runExecutorLauncher(): Unit = {
-    val hostname = Utils.localHostName
-    val amCores = sparkConf.get(AM_CORES)
-    rpcEnv = RpcEnv.create("sparkYarnAM", hostname, hostname, -1, sparkConf, securityMgr,
-      amCores, true)
-
-    // The client-mode AM doesn't listen for incoming connections, so report an invalid port.
-    registerAM(hostname, -1, sparkConf, sparkConf.getOption("spark.driver.appUIAddress"))
-
-    // The driver should be up and listening, so unlike cluster mode, just try to connect to it
-    // with no waiting or retrying.
-    val (driverHost, driverPort) = Utils.parseHostPort(args.userArgs(0))
-    val driverRef = rpcEnv.setupEndpointRef(
-      RpcAddress(driverHost, driverPort),
-      YarnSchedulerBackend.ENDPOINT_NAME)
+    var driverRef : RpcEndpointRef = null
+    if (sparkConf.get(YARN_UNMANAGED_AM)) {
```

Comment:
I'm not a big fan of this change. Feels like you should have a different method here called `runUnmanaged` that is called instead of `run()`, and takes an `RpcEnv`.

That way you don't need to keep `clientRpcEnv` at all since it would be local to that method, since nothing else here needs it. In fact even `rpcEnv` could go away and become a parameter to `createAllocator`...

Reply:
Refactored to a method runUnmanaged, please let me know if anything can be done better.